srml/grandpa/src/lib.rs

```diff
@@ -376,3 +378,53 @@ impl<T: Trait> finality_tracker::OnFinalizationStalled<T::BlockNumber> for Modul
 		<Stalled<T>>::put((further_wait, median));
 	}
 }
+
+// TODO [slashing]: Integrate this.
+/// A grandpa equivocation offence report.
+#[allow(dead_code)]
+struct GrandpaEquivocationOffence<FullIdentification> {
+	/// A round in which the incident happened.
+	round: u64,
+	/// The session index that starts an era in which the incident happened.
+	current_era_start_session_index: u32, // TODO [slashing]: Should be a SessionIndex.
+	/// The size of the validator set at the time of the offence.
+	validators_count: u32,
+	/// The authority which produced this equivocation.
+	offender: FullIdentification,
+}
+
+impl<FullIdentification: Clone> Offence<FullIdentification> for GrandpaEquivocationOffence<FullIdentification> {
+	const ID: Kind = *b"grandpa:equivoca";
+
+	fn offenders(&self) -> Vec<FullIdentification> {
+		vec![self.offender.clone()]
+	}
+
+	fn current_era_start_session_index(&self) -> u32 { // TODO [slashing]: Should be a SessionIndex.
+		self.current_era_start_session_index
+	}
+
+	fn validators_count(&self) -> u32 {
+		self.validators_count
+	}
+
+	fn time_slot(&self) -> TimeSlot {
+		self.round as TimeSlot
+	}
+
+	fn slash_fraction(
+		offenders_count: u32,
+		validators_count: u32,
+	) -> Perbill {
+		// the formula is min((3k / n)^2, 1)
+		let x = Perbill::from_rational_approximation(3 * offenders_count, validators_count);
+
+		// _ ^ 2
+		// For now, Perbill doesn't support taking the power of it. Until it does, do it manually.
+		// The conversion to u64 is performed to guarantee it fits.
+		// TODO: #3189 should fix this.
+		let x = x.into_parts();
+		let x = ((x as u64 * x as u64) / 1_000_000_000) as u32;
```

Comment:
is it actually squaring the denominator correctly?

Reply:
Good point, fixed.